edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/ModuleClientProvider.cs

```diff
@@ -2,30 +2,248 @@
+            IModuleClient moduleClient = new ModuleClient(sdkModuleClient, this.idleTimeout, this.closeOnIdleTimeout);
+            return moduleClient;
         }
 
-        public Task<IModuleClient> Create(
-            ConnectionStatusChangesHandler statusChangedHandler,
-            Func<IModuleClient, Task> initialize) =>
-            ModuleClient.Create(Option.Some(this.edgeAgentConnectionString), this.upstreamProtocol, statusChangedHandler, initialize, this.proxy, this.productInfo);
+        static ITransportSettings GetTransportSettings(UpstreamProtocol protocol, Option<IWebProxy> proxy)
+        {
+            switch (protocol)
+            {
+                case UpstreamProtocol.Amqp:
+                {
+                    var settings = new AmqpTransportSettings(TransportType.Amqp_Tcp_Only);
+                    proxy.ForEach(p => settings.Proxy = p);
+                    return settings;
+                }
+
+                case UpstreamProtocol.AmqpWs:
+                {
+                    var settings = new AmqpTransportSettings(TransportType.Amqp_WebSocket_Only);
+                    proxy.ForEach(p => settings.Proxy = p);
+                    return settings;
+                }
+
+                case UpstreamProtocol.Mqtt:
+                {
+                    var settings = new MqttTransportSettings(TransportType.Mqtt_Tcp_Only);
+                    proxy.ForEach(p => settings.Proxy = p);
+                    return settings;
+                }
+
+                case UpstreamProtocol.MqttWs:
+                {
+                    var settings = new MqttTransportSettings(TransportType.Mqtt_WebSocket_Only);
+                    proxy.ForEach(p => settings.Proxy = p);
+                    return settings;
+                }
+
+                default:
+                {
+                    throw new InvalidEnumArgumentException();
+                }
+            }
+        }
+
+        static Task<T> ExecuteWithRetry<T>(Func<Task<T>> func, Action<RetryingEventArgs> onRetry)
+        {
+            var transientRetryPolicy = new RetryPolicy(TransientErrorDetectionStrategy, TransientRetryStrategy);
+            transientRetryPolicy.Retrying += (_, args) => onRetry(args);
+            return transientRetryPolicy.ExecuteAsync(func);
+        }
+
+        async Task<ISdkModuleClient> CreateSdkModuleClientWithRetry(ConnectionStatusChangesHandler statusChangedHandler)
+        {
+            try
+            {
+                ISdkModuleClient moduleClient = await ExecuteWithRetry(
+                    () => this.CreateSdkModuleClient(statusChangedHandler),
+                    Events.RetryingDeviceClientConnection);
+                Events.DeviceClientCreated();
+                return moduleClient;
+            }
+            catch (Exception e)
+            {
+                Events.DeviceClientSetupFailed(e);
+                Environment.Exit(1);
```

Comment:
we just exit if we can get a client after the retry schedule? The default is infinite retries

Reply:
No, we exit if ExecuteWithRetry throws (which it will only for really bad errors that we cannot handle). That basically means that the EdgeAgent cannot connect to IoTHub anymore, so it exits. 
ExecuteWithRetry will do infinite retries.